java/AxivityParser.java

```diff
@@ -0,0 +1,373 @@
+                short yRaw = 0;
+                short zRaw = 0;
+                double x = 0.0;
+                double y = 0.0;
+                double z = 0.0;
+
+                // loop through each line in data block and check if it is last
+                // in epoch, then write epoch summary to file.
+                // An epoch will have a start+end time, and fixed duration
+                for (int i = 0; i < sampleCount; i++) {
+                    if (USE_PRECISE_TIME) {
+                        // Calculate each sample's time, not successively adding
+                        // so that we don't accumulate any errors
+                        blockTime = firstSampleTime.plusNanos((long) (i * (double) spanNanos / sampleCount));
+                    } else if (i == 0) {
+                            blockTime = firstSampleTime; // emulate original behaviour
+                    } else {
+                            blockTime = blockTime.plusNanos(secs2Nanos(1.0 / sampleFreq));
+                    }
+
+                    if (bytesPerSample == 4) {
+                        try {
+                            value = getUnsignedInt(buf, 30 + 4 * i);
+                        } catch (Exception excep) {
+                            errCounter += 1;
+                            System.err.println("xyz reading err: " + excep.toString());
+                            break; // rest of block/page may be corrupted
+                        }
+                        // Sign-extend 10-bit values, adjust for exponents
+                        xRaw = (short) ((short) (0xffffffc0 & (value << 6)) >> (6 - ((value >> 30) & 0x03)));
+                        yRaw = (short) ((short) (0xffffffc0 & (value >> 4)) >> (6 - ((value >> 30) & 0x03)));
+                        zRaw = (short) ((short) (0xffffffc0 & (value >> 14)) >> (6 - ((value >> 30) & 0x03)));
+                    } else if (bytesPerSample == 6) {
+                        try {
+                            xRaw = buf.getShort(30 + 2 * NUM_AXES_PER_SAMPLE * i + 0);
+                            yRaw = buf.getShort(30 + 2 * NUM_AXES_PER_SAMPLE * i + 2);
+                            zRaw = buf.getShort(30 + 2 * NUM_AXES_PER_SAMPLE * i + 4);
+                        } catch (Exception excep) {
+                            errCounter += 1;
+                            System.err.println("xyz read err: " + excep.toString());
+                            break; // rest of block/page may be corrupted
+                        }
+                    } else {
+                        xRaw = 0;
+                        yRaw = 0;
+                        zRaw = 0;
+                    }
+                    x = xRaw / 256.0;
+                    y = yRaw / 256.0;
+                    z = zRaw / 256.0;
+                    t = zonedWithDSTCorrection(blockTime).toInstant().toEpochMilli();
+
+                    try {
+                        writer.writeData(t, (float) x, (float) y, (float) z);
```

Comment:
Would it be better practice to have writer passed in as an argument to this block or code (rather than being used as a global variable)?

Reply:
I thought about this and concluded that it'd be best to use an object-oriented pattern and make `writer` a private attribute of a parser object. In this view, a `writer` is an inherent part of a parser instance. This also saves us from having to pass `writer` to `parseCwaBuffer`, in fact, `writer` doesn't change at all during the loop and is only passed so that the function gets a hold of the pointer. In the process, I also made `errCounter`, `lastBlockTimeIndex`, and `lastBlockTime` private attributes of the parser object for similar reasons:

#### Current
```java
// AxivityReader.java

// Lines 36-66
// Although errCounter in practice acts as a simple int counter, it has to be defined as 
// a size 1 array so that we can pass it to readAccBuffer below and persist changes. 
// Same goes for lastBlockTime and lastBlockTimeIndex.
int[] errCounter = new int[] { 0 };
LocalDateTime[] lastBlockTime = { null };
int[] lastBlockTimeIndex = { 0 };

// ...
try {
    // ...
    while (rawAccReader.read(buf) != -1) {
        // Here the only thing that's worth passing is buf
        // The other variables are mostly bookkeeping variables
        readCwaBuffer(buf, USE_PRECISE_TIME, lastBlockTime, lastBlockTimeIndex, header, errCounter, epochWriter);
        buf.clear();
    }
} catch {
// ...
}

// Further down, inside readCwaBuffer()

// Because we had to pass errCounter, lastBlockTime, lastBlockTimeIndex as size 1 arrays,
// we get some strange looking lines for simple increments, assignments, comparisons:

// Line 240
if (lastBlockTime[0] != null) {
// ...
}

// Line 259
lastBlockTime[0] = blockTime;
lastBlockTimeIndex[0] = timestampOffset - sampleCount;

// Line 294
errCounter[0] += 1;
```
#### Proposed
```java
// AxivityParser.java -- this PR

// Using an object-oriented pattern, keep errCounter, lastBlockTime, lastBlockTimeIndex and writer as
// private variables of "this" object
private int errCounter = 0;
private LocalDateTime lastBlockTime = null;
private int lastBlockTimeIndex = 0;
private NpyWriter writer;

// In parse()

try {
    // ...
    while (accChannel.read(buf) != -1) {
        // Just passing buf
        parseCwaBuffer(buf);
        buf.clear();
    }
} catch {
// ...

// Further down, in parseCwaBuffer()
// Variables used intuitively
if (lastBlockTime[0] != null) {
// ...
}
// ...
lastBlockTime = blockTime;
lastBlockTimeIndex = timestampOffset - sampleCount;
// ...
errCounter += 1;
```

Reply:
Looking at ActigraphReader and GENEActiveReader, I'm considering reverting to the object-less, static-based pattern as otherwise it'll take too much time for little gain.

Reply:
That sounds fine.